### ISSUE TITLE
[DFC Products] Enable Variant Product mappings across DFC endpoints

### ISSUE THREAD

**Issue by RaggedStaff** [openfoodfoundation/openfoodnetwork#12984]
#### :information_source: Funded Feature. Please track ALL ASSOCIATED WORK under the associated tracking code `#11678 DFC Orders`

We are enhancing DFC to include the ability to track Product > Variant relationships.

2 new relationships will be added to `DefinedProduct` : `hasVariant` (indicates the Product has a Variant with the given Id) & `isVariantOf` (indicates the Product is a Variant of the Product with the given Id).

Additional objects will be incorporated in a future version of the Ontology to enhance this further. Probably `Variant` as a subclass of `DefinedProduct`, details TBC. For now, we're just implementing these 2 relationships to hold this data. 

Both Shopify & OFN support the concept of Product/Variants so hopefully this should be fairly seamless. We're not restricting anything from either Product or Variant at the moment, although we might want to look at the [DFC issue](https://github.com/datafoodconsortium/ontology/issues/150#issuecomment-2419774942) for guidelines as to what to include/not include.

I'm thinking we need to amend Product import & also export. WDYT @mkllnk ?

**Comment by mkllnk** [comment]
> we need to amend Product import & also export.

Yes, the export doesn't include master products at the moment. The import needs to recognise master products.

We'll need the new attributes in the data model for the connector.

**Comment by mkllnk** [comment]
Estimate 5 days.

* Update the connector.
* Include Spree::Product in the export.
* Add transformations in the export.
* Recognise master products in the import.
* Connect variants to Spree::Product.

**Comment by mkllnk** [comment]
This work will deprecate an existing property which we should communicate on release, maybe like this:


> There's an upcoming change to the DFC API and it may affect integrations. In the past, the DFC didn't support the product-variant relationship. As a workaround, we added a property called `ofn:spree_product_id` to create that relationship. This is now deprecated and will be removed at some point in the future.
> 
> The DFC now supports variants and there will be a new property `dfc-b:isVariantOf` which will have the same role. The value will be a URL containing the product id instead of just the plain product id. So any integrations using the old property need updating.

**Comment by mkllnk** [comment]
@RaggedStaff, I need some ideas for one of the tasks:

> Add transformations in the export.

Initially I thought this simple because we got variants with size data already stored in OFN. For example, we got olive oil in different bottles or cans:

* 500 ml
* 1 l
* 4 l
* 20 l

They are not easily transformed. A 20 liter can is not made out of 20 bottles with 1 liter each. A buying group may still order a 20 liter can and divide it up in smaller bottles for the members of the group. With our current logic of mapping between retail and wholesale, I could apply a logic of adding transformations from small to big.

There are still a lot of options though:

* Do we add all possible transformations? That's 12 transformation flows for 4 variants.
* Do we just map from small to big? That's 6 transformations.
* Do we choose the biggest variant as the wholesale variant? Then we have only three transformations for each smaller variant into the big one. But you could argue that 4 liters is also wholesale, not just 20 liter.
* Do we not bother generating transformations and explore a new UI for this?

But now I also realise that this point is not in the issue description here. Did I make it up? Or did we discuss that before variants came along?

I think I should restrict this task to variants only and we discuss transformations separately.
